Fix indexing inside SplineX2X when outputing orbitals fewer than it holds

camelto2 opened the pull request:
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Found a segfault when trying to use hybridrep in orbital optimization. The source of the problem is that when doing orbital optimization, the underlying orbital set includes unoccupied orbitals. However, if we are calling the evaluations from inside a DiracDeterminant, the ValueVector psi size is ways set to the number of occupied orbitals, not the full size of the SPOSet. The fix here allows for proper indexing into psi, and this is actually producing a successful hyrbid rep optimization. 

The spline classes do not properly guard against accessing out of bounds for the ValueVector psi, and I expect this is more general problem for all SPOSets. 

TO DO:
- [x] Fix SplineR2R
- [x] Fix SplineC2C
- [x] add useful unit tests
- [x] Fix SplineC2R 
- [x] Fix SplineC2COMPTarget
- [x] Fix SplineC2ROMPTarget

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
M1 mac

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

camelto2 commented:
The current units tests that are failing, are actually failing as expected. This is due to other spline implementations needing the same fixes. I'll work on those next and get the test passing

camelto2 commented:
As is, the unit tests are modified to allow for the (requested number of orbitals) < (the number of orbitals defined in the SPOSet). The current unit test in test_hybridrep only checks SplineC2C and SplineR2R. These modifications to the unit test fail with an out of bounds access without the changes to SplineC2C and SplineR2R. The various assign_x functions in the splines now guard against out of bounds access for the requested number of orbitals, and then the tests pass

The diamond2x1x1 test can be modified in the same way to test SplineC2R, but I'm still taking a look at that. For now, I think we could just merge this, and do subsequent PRs on SplineC2R and SplineC2ROMPTarget/SplineC2COMPTarget if those need it

ye-luo commented on src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp:
```diff
@@ -167,7 +167,8 @@ inline void SplineC2C<ST>::assign_v(const PointType& r,
   // protect last
-  last = last > kPoints.size() ? kPoints.size() : last;
+  const size_t last_real = std::min(kPoints.size(), psi.size());
+  last = last > last_real ? last_real : last;
```
Should be called last_cplx as the last complex output orbital.

camelto2 replied:
done

ye-luo commented on src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp:
```diff
@@ -418,7 +421,8 @@ void SplineC2C<ST>::assign_vgh(const PointType& r,
 {
   // protect last
-  last = last > kPoints.size() ? kPoints.size() : last;
+  const size_t last_real = std::min(kPoints.size(), psi.size());
```
last_cplx here.

camelto2 replied:
done

ye-luo commented on src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp:
```diff
@@ -342,7 +344,8 @@ inline void SplineC2C<ST>::assign_vgl_from_l(const PointType& r, ValueVector& ps
   const ST* restrict g2 = myG.data(2);
 
-  const size_t N = last_spo - first_spo;
+  const size_t last_real = last_spo > psi.size() ? psi.size() : last_spo;
```
last_cplx here.

camelto2 replied:
done

ye-luo commented on src/QMCWaveFunctions/BsplineFactory/SplineC2C.cpp:
```diff
@@ -256,7 +257,8 @@ inline void SplineC2C<ST>::assign_vgl(const PointType& r,
 {
   // protect last
-  last = last > kPoints.size() ? kPoints.size() : last;
+  const int last_real = std::min(kPoints.size(), psi.size());
```
last_cplx here.

camelto2 replied:
done

prckent commented:
Test this please

camelto2 commented:
Well the V100 Clang offload tests were failing, which makes me think the tests I modified actually end up going through the SplineC2COMPTarget, so i need to include those fixes in this PR as well. So I went ahead and fixed SplineC2R and modified a test for that as well since it seems all the Spline classes need a fix

camelto2 commented:
Ok, I think all tests should pass now

ye-luo commented:
Will approve and merge once the CI update is merged.

ye-luo commented:
Test this please
